lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -32,7 +32,10 @@ def result
         # continue by "finishing" the value, eg. executing sub-fields or coercing values
         def get_finished_value(raw_value)
           raise raw_value if raw_value.instance_of?(GraphQL::ExecutionError)
-          return nil if raw_value.nil?
+          if raw_value.nil?
+            raise GraphQL::ExecutionError, "Type mismatch resolving '#{field.name}'" if field.type.kind.non_null?
```

Comment:
We should probably add a couple type kind strategies in ValueResolution for interface and union types, and let them resolve the type, since I think that is unnecessary in the other cases.

I think we should move the `nil` checking into BaseResolution#result, which should just return `nil` if the value is `nil`, otherwise, it should call `coerce_result`.  The classes that derive from BaseResolution should rename their `result` method to `coerce_result`.  We can then override `result` in NonNullResolution to raise if the value is `nil`.

I'm not exactly sure how a scalar type should specify what should be done for a `nil` on non-null type, but NonNullResolution#result could call that `if value.nil? && field_type.of_type == GraphQL::TypeKinds::SCALAR`.